File: Packages/TabCore/Sources/TabCore/SplitCalculator.swift
import Foundation

public enum SplitCalculatorError: Error, Equatable, Sendable {
    case emptyParticipants
    case duplicateParticipant(UUID)
    case unsupportedSplitType(SplitType)
    case exactAmountsRequired
    case missingAmountForParticipant(UUID)
    case extraAmountForNonParticipant(UUID)
    case amountHasTooManyFractionDigits(currency: String, maximumFractionDigits: Int)
    case amountsDoNotSumToTotal(expected: Decimal, actual: Decimal)
    case sharesRequired
    case missingShareForParticipant(UUID)
    case extraShareForNonParticipant(UUID)
    case nonPositiveShare(UUID)
    case percentagesRequired
    case missingPercentageForParticipant(UUID)
    case extraPercentageForNonParticipant(UUID)
    case nonPositivePercentage(UUID)
    case percentagesDoNotSumTo100(actual: Decimal)
}

public enum SplitCalculator {
    public static func calculate(
        totalAmount: Decimal,
        currency: String,
        participants: [UUID],
        splitType: SplitType,
        exactAmounts: [UUID: Decimal]? = nil,
        shares: [UUID: Decimal]? = nil,
        percentages: [UUID: Decimal]? = nil
    ) throws -> [ExpenseSplit] {
        guard !participants.isEmpty else {
            throw SplitCalculatorError.emptyParticipants
        }

        var seen = Set<UUID>()
        for participant in participants where !seen.insert(participant).inserted {
            throw SplitCalculatorError.duplicateParticipant(participant)
        }

        try validatePrecision(totalAmount, currency: currency)

        switch splitType {
        case .equal:
            return calculateEqual(total: totalAmount, currency: currency, participants: participants)
        case .exact:
            guard let amounts = exactAmounts else {
                throw SplitCalculatorError.exactAmountsRequired
            }
            return try calculateExact(total: totalAmount, currency: currency, participants: participants, amounts: amounts)
        case .shares:
            guard let shares else {
                throw SplitCalculatorError.sharesRequired
            }
            return try calculateShares(total: totalAmount, currency: currency, participants: participants, shares: shares)
        case .percentage:
            guard let percentages else {
                throw SplitCalculatorError.percentagesRequired
            }
            return try calculatePercentages(total: totalAmount, currency: currency, participants: participants, percentages: percentages)
        case .adjustment:
            throw SplitCalculatorError.unsupportedSplitType(splitType)
        }
    }

    /// Equal percentages at 2 fraction digits that sum to exactly 100; the
    /// leftover basis points go to the lexicographically lowest UUIDs (the
    /// same convention as equal-split remainders). Used to seed percentage UI.
    public static func equalPercentages(participants: [UUID]) -> [UUID: Decimal] {
        guard !participants.isEmpty else { return [:] }
        let totalBasisPoints = 10_000
        let base = totalBasisPoints / participants.count
        let remainder = totalBasisPoints % participants.count
        let sorted = participants.sorted { $0.uuidString < $1.uuidString }
        var result: [UUID: Decimal] = [:]
        for (index, id) in sorted.enumerated() {
            result[id] = Decimal(base + (index < remainder ? 1 : 0)) / 100
        }
        return result
    }

    // Distributes `total` evenly at the smallest supported unit for the currency.
    // Any remainder is assigned one minor unit at a time to the lowest sorted UUIDs.
    private static func calculateEqual(total: Decimal, currency: String, participants: [UUID]) -> [ExpenseSplit] {
        let n = Decimal(participants.count)
        let multiplier = CurrencyCatalog.minorUnitMultiplier(for: currency)
        let totalMinorUnits = roundToInteger(total * multiplier)
        let baseMinorUnits = roundDownToInteger(totalMinorUnits / n)
        let baseShare = baseMinorUnits / multiplier
        let remainderUnits = totalMinorUnits - baseMinorUnits * n

        let extraCount = (remainderUnits as NSDecimalNumber).intValue
        let sortedIDs = participants.sorted { $0.uuidString < $1.uuidString }
        let bonusIDs = Set(sortedIDs.prefix(extraCount))

        let smallestUnit = Decimal(1) / multiplier
        return participants.map { id in
            let owed = bonusIDs.contains(id) ? (baseShare + smallestUnit) : baseShare
            return ExpenseSplit(participantID: id, amountOwed: owed, splitType: .equal)
        }
    }

    private static func calculateExact(
        total: Decimal,
        currency: String,
        participants: [UUID],
        amounts: [UUID: Decimal]
    ) throws -> [ExpenseSplit] {
        let participantSet = Set(participants)

        for key in amounts.keys where !participantSet.contains(key) {
            throw SplitCalculatorError.extraAmountForNonParticipant(key)
        }
        for participant in participants where amounts[participant] == nil {
            throw SplitCalculatorError.missingAmountForParticipant(participant)
        }

        for amount in amounts.values {
            try validatePrecision(amount, currency: currency)
        }

        let sum = amounts.values.reduce(Decimal(0), +)
        if sum != total {
            throw SplitCalculatorError.amountsDoNotSumToTotal(expected: total, actual: sum)
        }

        return participants.map { id in
            ExpenseSplit(participantID: id, amountOwed: amounts[id]!, splitType: .exact)
        }
    }

    private static func calculateShares(
        total: Decimal,
        currency: String,
        participants: [UUID],
        shares: [UUID: Decimal]
    ) throws -> [ExpenseSplit] {
        let participantSet = Set(participants)

        for key in shares.keys where !participantSet.contains(key) {
            throw SplitCalculatorError.extraShareForNonParticipant(key)
        }
        for participant in participants where shares[participant] == nil {
            throw SplitCalculatorError.missingShareForParticipant(participant)
        }
        for participant in participants.sorted(by: { $0.uuidString < $1.uuidString }) where shares[participant]! <= 0 {
            throw SplitCalculatorError.nonPositiveShare(participant)
        }

        let amounts = allocateProportionally(total: total, currency: currency, participants: participants, weights: shares)
        return participants.map { id in
            ExpenseSplit(participantID: id, amountOwed: amounts[id]!, splitType: .shares, shareUnits: shares[id])
        }
    }

    private static func calculatePercentages(
        total: Decimal,
        currency: String,
        participants: [UUID],
        percentages: [UUID: Decimal]
    ) throws -> [ExpenseSplit] {
        let participantSet = Set(participants)

        for key in percentages.keys where !participantSet.contains(key) {
            throw SplitCalculatorError.extraPercentageForNonParticipant(key)
        }
        for participant in participants where percentages[participant] == nil {
            throw SplitCalculatorError.missingPercentageForParticipant(participant)
        }
        for participant in participants.sorted(by: { $0.uuidString < $1.uuidString }) where percentages[participant]! <= 0 {
            throw SplitCalculatorError.nonPositivePercentage(participant)
        }

        let sum = participants.reduce(Decimal(0)) { $0 + percentages[$1]! }
        if sum != 100 {
            throw SplitCalculatorError.percentagesDoNotSumTo100(actual: sum)
        }

        let amounts = allocateProportionally(total: total, currency: currency, participants: participants, weights: percentages)
        return participants.map { id in
            ExpenseSplit(participantID: id, amountOwed: amounts[id]!, splitType: .percentage, percentage: percentages[id])
        }
    }

    // Weight-proportional allocation at the smallest supported unit for the
    // currency. Each participant gets floor(total * weight / weightSum) minor
    // units; the leftover units go out one at a time by largest fractional
    // remainder, tie-broken by lexicographically lowest UUID (deterministic,
    // like equal). Weights must be positive and non-empty (callers validate).
    private static func allocateProportionally(
        total: Decimal,
        currency: String,
        participants: [UUID],
        weights: [UUID: Decimal]
    ) -> [UUID: Decimal] {
        let weightSum = participants.reduce(Decimal(0)) { $0 + weights[$1]! }
        let multiplier = CurrencyCatalog.minorUnitMultiplier(for: currency)
        let totalMinorUnits = roundToInteger(total * multiplier)

        var floorMinorUnits: [UUID: Decimal] = [:]
        var fractionalParts: [UUID: Decimal] = [:]
        var floorSum = Decimal(0)
        for id in participants {
            let ideal = totalMinorUnits * weights[id]! / weightSum
            let floored = roundDownToInteger(ideal)
            floorMinorUnits[id] = floored
            fractionalParts[id] = ideal - floored
            floorSum += floored
        }

        // Largest fractional remainder first; UUID breaks ties.
        let order = participants.sorted { a, b in
            if fractionalParts[a]! != fractionalParts[b]! {
                return fractionalParts[a]! > fractionalParts[b]!
            }
            return a.uuidString < b.uuidString
        }

        var remaining = ((totalMinorUnits - floorSum) as NSDecimalNumber).intValue
        for id in order where remaining > 0 {
            floorMinorUnits[id]! += 1
            remaining -= 1
        }
        // Decimal division rounding at precision limits could in principle push
        // a floor one unit high; claw back so the splits always sum to total.
        for id in order.reversed() where remaining < 0 {
            if floorMinorUnits[id]! > 0 {
                floorMinorUnits[id]! -= 1
                remaining += 1
            }
        }

        return floorMinorUnits.mapValues { $0 / multiplier }
    }

    private static func validatePrecision(_ amount: Decimal, currency: String) throws {
        guard CurrencyCatalog.hasValidPrecision(amount, currency: currency) else {
            throw SplitCalculatorError.amountHasTooManyFractionDigits(
                currency: CurrencyCatalog.normalizedCode(currency),
                maximumFractionDigits: CurrencyCatalog.fractionDigits(for: currency)
            )
        }
    }

    private static func roundToInteger(_ value: Decimal) -> Decimal {
        var input = value
        var result = Decimal()
        NSDecimalRound(&result, &input, 0, .plain)
        return result
    }

    private static func roundDownToInteger(_ value: Decimal) -> Decimal {
        var input = value
        var result = Decimal()
        NSDecimalRound(&result, &input, 0, .down)
        return result
    }
}

File: Apps/TabAndroid/core/domain/src/main/kotlin/com/rithwikshetty/tab/domain/SplitCalculator.kt
package com.rithwikshetty.tab.domain

import java.math.BigDecimal
import java.math.RoundingMode
import java.util.UUID

public sealed class SplitCalculatorException(message: String) : IllegalArgumentException(message) {
    public data object EmptyParticipants : SplitCalculatorException("Participants cannot be empty.")
    public data class DuplicateParticipant(public val id: UUID) :
        SplitCalculatorException("Duplicate participant: $id")
    public data class UnsupportedSplitType(public val type: SplitType) :
        SplitCalculatorException("Unsupported split type: $type")
    public data object ExactAmountsRequired : SplitCalculatorException("Exact amounts are required.")
    public data class MissingAmount(public val id: UUID) :
        SplitCalculatorException("Missing amount for participant: $id")
    public data class ExtraAmount(public val id: UUID) :
        SplitCalculatorException("Amount supplied for non-participant: $id")
    public data class InvalidPrecision(public val currency: String, public val maximumFractionDigits: Int) :
        SplitCalculatorException("Amount exceeds $currency precision.")
    public data class AmountsDoNotSum(public val expected: BigDecimal, public val actual: BigDecimal) :
        SplitCalculatorException("Amounts do not sum to total.")
    public data object SharesRequired : SplitCalculatorException("Shares are required.")
    public data class MissingShare(public val id: UUID) :
        SplitCalculatorException("Missing share for participant: $id")
    public data class ExtraShare(public val id: UUID) :
        SplitCalculatorException("Share supplied for non-participant: $id")
    public data class NonPositiveShare(public val id: UUID) :
        SplitCalculatorException("Share must be positive: $id")
    public data object PercentagesRequired : SplitCalculatorException("Percentages are required.")
    public data class MissingPercentage(public val id: UUID) :
        SplitCalculatorException("Missing percentage for participant: $id")
    public data class ExtraPercentage(public val id: UUID) :
        SplitCalculatorException("Percentage supplied for non-participant: $id")
    public data class NonPositivePercentage(public val id: UUID) :
        SplitCalculatorException("Percentage must be positive: $id")
    public data class PercentagesDoNotSumTo100(public val actual: BigDecimal) :
        SplitCalculatorException("Percentages do not sum to 100.")
}

public object SplitCalculator {
    public fun calculate(
        totalAmount: BigDecimal,
        currency: String,
        participants: List<UUID>,
        splitType: SplitType,
        exactAmounts: Map<UUID, BigDecimal>? = null,
        shares: Map<UUID, BigDecimal>? = null,
        percentages: Map<UUID, BigDecimal>? = null,
    ): List<ExpenseSplit> {
        if (participants.isEmpty()) throw SplitCalculatorException.EmptyParticipants
        participants.groupingBy { it }.eachCount().entries.firstOrNull { it.value > 1 }?.let {
            throw SplitCalculatorException.DuplicateParticipant(it.key)
        }
        validatePrecision(totalAmount, currency)

        return when (splitType) {
            SplitType.EQUAL -> calculateEqual(totalAmount, currency, participants)
            SplitType.EXACT -> calculateExact(
                totalAmount,
                currency,
                participants,
                exactAmounts ?: throw SplitCalculatorException.ExactAmountsRequired,
            )
            SplitType.SHARES -> calculateWeighted(
                totalAmount,
                currency,
                participants,
                shares ?: throw SplitCalculatorException.SharesRequired,
                SplitType.SHARES,
            )
            SplitType.PERCENTAGE -> calculateWeighted(
                totalAmount,
                currency,
                participants,
                percentages ?: throw SplitCalculatorException.PercentagesRequired,
                SplitType.PERCENTAGE,
            )
            SplitType.ADJUSTMENT -> throw SplitCalculatorException.UnsupportedSplitType(splitType)
        }
    }

    public fun equalPercentages(participants: List<UUID>): Map<UUID, BigDecimal> {
        if (participants.isEmpty()) return emptyMap()
        val base = 10_000 / participants.size
        val remainder = 10_000 % participants.size
        return participants.sortedBy(UUID::toString).mapIndexed { index, id ->
            id to BigDecimal(base + if (index < remainder) 1 else 0).movePointLeft(2)
        }.toMap()
    }

    private fun calculateEqual(
        total: BigDecimal,
        currency: String,
        participants: List<UUID>,
    ): List<ExpenseSplit> {
        val multiplier = CurrencyCatalog.minorUnitMultiplier(currency)
        val totalMinor = total.multiply(multiplier).setScale(0, RoundingMode.HALF_UP).toBigIntegerExact()
        val count = participants.size.toBigInteger()
        val base = totalMinor.divide(count)
        val remainder = totalMinor.remainder(count).toInt()
        val bonusIds = participants.sortedBy(UUID::toString).take(remainder).toSet()
        return participants.map { id ->
            val minor = base + if (id in bonusIds) java.math.BigInteger.ONE else java.math.BigInteger.ZERO
            ExpenseSplit(id, BigDecimal(minor).divide(multiplier), SplitType.EQUAL)
        }
    }

    private fun calculateExact(
        total: BigDecimal,
        currency: String,
        participants: List<UUID>,
        amounts: Map<UUID, BigDecimal>,
    ): List<ExpenseSplit> {
        val participantSet = participants.toSet()
        amounts.keys.firstOrNull { it !in participantSet }?.let {
            throw SplitCalculatorException.ExtraAmount(it)
        }
        participants.firstOrNull { it !in amounts }?.let {
            throw SplitCalculatorException.MissingAmount(it)
        }
        amounts.values.forEach { validatePrecision(it, currency) }
        val sum = amounts.values.fold(BigDecimal.ZERO, BigDecimal::add)
        if (sum.compareTo(total) != 0) throw SplitCalculatorException.AmountsDoNotSum(total, sum)
        return participants.map { ExpenseSplit(it, amounts.getValue(it), SplitType.EXACT) }
    }

    private fun calculateWeighted(
        total: BigDecimal,
        currency: String,
        participants: List<UUID>,
        weights: Map<UUID, BigDecimal>,
        type: SplitType,
    ): List<ExpenseSplit> {
        val participantSet = participants.toSet()
        weights.keys.firstOrNull { it !in participantSet }?.let {
            if (type == SplitType.SHARES) throw SplitCalculatorException.ExtraShare(it)
            throw SplitCalculatorException.ExtraPercentage(it)
        }
        participants.firstOrNull { it !in weights }?.let {
            if (type == SplitType.SHARES) throw SplitCalculatorException.MissingShare(it)
            throw SplitCalculatorException.MissingPercentage(it)
        }
        participants.sortedBy(UUID::toString).firstOrNull {
            weights.getValue(it).compareTo(BigDecimal.ZERO) <= 0
        }?.let {
            if (type == SplitType.SHARES) throw SplitCalculatorException.NonPositiveShare(it)
            throw SplitCalculatorException.NonPositivePercentage(it)
        }
        if (type == SplitType.PERCENTAGE) {
            val sum = participants.fold(BigDecimal.ZERO) { value, id -> value + weights.getValue(id) }
            if (sum.compareTo(BigDecimal("100")) != 0) {
                throw SplitCalculatorException.PercentagesDoNotSumTo100(sum)
            }
        }
        val amounts = allocateProportionally(total, currency, participants, weights)
        return participants.map { id ->
            ExpenseSplit(
                participantId = id,
                amountOwed = amounts.getValue(id),
                splitType = type,
                shareUnits = weights[id].takeIf { type == SplitType.SHARES },
                percentage = weights[id].takeIf { type == SplitType.PERCENTAGE },
            )
        }
    }

    private fun allocateProportionally(
        total: BigDecimal,
        currency: String,
        participants: List<UUID>,
        weights: Map<UUID, BigDecimal>,
    ): Map<UUID, BigDecimal> {
        val multiplier = CurrencyCatalog.minorUnitMultiplier(currency)
        val totalMinor = total.multiply(multiplier).setScale(0, RoundingMode.HALF_UP)
        val weightSum = participants.fold(BigDecimal.ZERO) { value, id -> value + weights.getValue(id) }
        val floors = mutableMapOf<UUID, BigDecimal>()
        val fractions = mutableMapOf<UUID, BigDecimal>()
        participants.forEach { id ->
            val ideal = totalMinor.multiply(weights.getValue(id)).divide(weightSum, 24, RoundingMode.HALF_UP)
            val floor = ideal.setScale(0, RoundingMode.DOWN)
            floors[id] = floor
            fractions[id] = ideal - floor
        }
        var remaining = totalMinor.subtract(floors.values.fold(BigDecimal.ZERO, BigDecimal::add)).intValueExact()
        val order = participants.sortedWith(
            compareByDescending<UUID> { fractions.getValue(it) }.thenBy(UUID::toString),
        )
        for (id in order) {
            if (remaining <= 0) break
            floors[id] = floors.getValue(id) + BigDecimal.ONE
            remaining--
        }
        for (id in order.asReversed()) {
            if (remaining >= 0) break
            if (floors.getValue(id) > BigDecimal.ZERO) {
                floors[id] = floors.getValue(id) - BigDecimal.ONE
                remaining++
            }
        }
        return floors.mapValues { (_, minor) -> minor.divide(multiplier) }
    }

    private fun validatePrecision(amount: BigDecimal, currency: String) {
        if (!CurrencyCatalog.hasValidPrecision(amount, currency)) {
            throw SplitCalculatorException.InvalidPrecision(
                CurrencyCatalog.normalizedCode(currency),
                CurrencyCatalog.fractionDigits(currency),
            )
        }
    }
}
